File: app/src/main/java/com/spotik/app/ui/viewmodel/ProfileViewModel.kt
package com.spotik.app.ui.viewmodel

import android.net.Uri
import androidx.compose.runtime.getValue
import androidx.compose.runtime.mutableIntStateOf
import androidx.compose.runtime.mutableStateOf
import androidx.compose.runtime.setValue
import androidx.lifecycle.ViewModel
import androidx.lifecycle.viewModelScope
import com.spotik.app.data.api.RetrofitClient
import com.spotik.app.data.api.UpdateProfileRequest
import com.spotik.app.data.auth.AuthManager
import kotlinx.coroutines.launch

/**
 * ViewModel for the Profile screen.
 * Loads data from the server (with local fallback from AuthManager),
 * and syncs edits back.
 */
class ProfileViewModel : ViewModel() {

    var name by mutableStateOf("")
        private set
    var age by mutableIntStateOf(18)
        private set
    var bio by mutableStateOf("")
        private set
    var city by mutableStateOf("")
        private set
    var avatarUri by mutableStateOf<Uri?>(null)
        private set

    var isLoading by mutableStateOf(true)
        private set

    /* ── Editable copies (used inside Edit screen) ── */
    var editName by mutableStateOf("")
    var editAge by mutableIntStateOf(18)
    var editBio by mutableStateOf("")
    var editAvatarUri by mutableStateOf<Uri?>(null)

    init {
        loadProfile()
    }

    /** Start editing — copy current values into edit fields. */
    fun beginEdit() {
        editName = name
        editAge = age
        editBio = bio
        editAvatarUri = avatarUri
    }

    /** Apply edits to local state and sync to server. */
    fun applyEdit() {
        name = editName
        age = editAge
        bio = editBio
        avatarUri = editAvatarUri

        // Save locally for offline
        AuthManager.saveProfile(
            name = name,
            age = age,
            bio = bio,
            city = city,
            avatarUrl = avatarUri?.toString(),
        )

        // Push to server
        viewModelScope.launch {
            try {
                RetrofitClient.api.updateProfile(
                    UpdateProfileRequest(
                        name = name,
                        age = age,
                        bio = bio,
                        city = city,
                        avatarUrl = avatarUri?.toString(),
                    )
                )
            } catch (_: Exception) {
                // Will sync next time
            }
        }
    }

    fun setEditAvatar(uri: Uri?) {
        editAvatarUri = uri
    }

    /** Load profile: first from local prefs, then overwrite with server data. */
    private fun loadProfile() {
        // Immediate local data
        name = AuthManager.userName ?: ""
        age = AuthManager.userAge
        bio = AuthManager.userBio ?: ""
        city = AuthManager.userCity ?: ""
        val savedAvatar = AuthManager.userAvatar
        if (!savedAvatar.isNullOrBlank()) {
            avatarUri = Uri.parse(savedAvatar)
        }
        isLoading = false

        // Then fetch from server
        if (AuthManager.token != null) {
            viewModelScope.launch {
                try {
                    val profile = RetrofitClient.api.getProfile()
                    name = profile.name
                    age = profile.age
                    bio = profile.bio ?: ""
                    city = profile.city
                    if (!profile.avatarUrl.isNullOrBlank()) {
                        avatarUri = Uri.parse(profile.avatarUrl)
                    }
                    // Update local cache
                    AuthManager.saveProfile(
                        name = profile.name,
                        age = profile.age,
                        bio = profile.bio,
                        city = profile.city,
                        avatarUrl = profile.avatarUrl,
                    )
                } catch (_: Exception) {
                    // Server unavailable — local data is already set
                }
            }
        }
    }

    /** Force reload (e.g. after login). */
    fun refresh() {
        loadProfile()
    }
}

File: ios/Sources/ViewModels/ProfileViewModel.swift
import SwiftUI

@MainActor
final class ProfileViewModel: ObservableObject {
    @Published var profile: UserProfile? = nil
    @Published var isLoading = false
    @Published var isEditing = false
    @Published var showLogoutDialog = false

    // Edit fields
    @Published var editName: String = ""
    @Published var editAge: Int = 18
    @Published var editBio: String = ""
    @Published var editCity: String = ""

    // Stats (mock for demo)
    @Published var likesCount: Int = 0
    @Published var viewsCount: Int = 0
    @Published var referralsCount: Int = 0

    // Activity
    @Published var activityData: [Double] = [0.3, 0.5, 0.8, 0.4, 0.9, 0.6, 0.7]

    func loadProfile() {
        // Use cached data first for instant display
        if let name = AuthManager.shared.userName {
            profile = UserProfile(
                id: AuthManager.shared.userId ?? "",
                name: name,
                age: AuthManager.shared.userAge,
                city: AuthManager.shared.userCity ?? "",
                bio: AuthManager.shared.userBio,
                avatarUrl: AuthManager.shared.userAvatar,
                email: AuthManager.shared.email
            )
            editName = name
            editAge = AuthManager.shared.userAge
            editBio = AuthManager.shared.userBio ?? ""
            editCity = AuthManager.shared.userCity ?? ""
        }

        // Then refresh from server
        Task {
            do {
                let p = try await APIService.shared.getProfile()
                profile = p
                editName = p.name
                editAge = p.age
                editBio = p.bio ?? ""
                editCity = p.city
                AuthManager.shared.saveProfile(
                    name: p.name, age: p.age, bio: p.bio,
                    city: p.city, avatarUrl: p.avatarUrl
                )
                // Mock stats — replace with real API when available
                likesCount   = Int.random(in: 120...980)
                viewsCount   = Int.random(in: 2000...12000)
                referralsCount = Int.random(in: 3...42)
            } catch {
                // Keep cached version
            }
        }
    }

    func saveEdits() {
        Task {
            let body = UpdateProfileRequest(
                name: editName.isEmpty ? nil : editName,
                age: editAge,
                bio: editBio.isEmpty ? nil : editBio,
                city: editCity.isEmpty ? nil : editCity,
                avatarUrl: nil
            )
            do {
                _ = try await APIService.shared.updateProfile(body)
                AuthManager.shared.saveProfile(
                    name: editName, age: editAge, bio: editBio,
                    city: editCity, avatarUrl: profile?.avatarUrl
                )
                profile = UserProfile(
                    id: profile?.id ?? "",
                    name: editName, age: editAge,
                    city: editCity, bio: editBio,
                    avatarUrl: profile?.avatarUrl,
                    email: profile?.email
                )
                isEditing = false
            } catch {
                // silent fail — local update already done
                isEditing = false
            }
        }
    }

    func logout(onLogout: @escaping () -> Void) {
        Task {
            _ = try? await APIService.shared.logout()
            AuthManager.shared.logout()
            onLogout()
        }
    }
}
